rfcs/20200804-primitive-fallback/README.md

```diff
@@ -0,0 +1,172 @@
+
+In this case, the control flow can be simplified to:
+
+~~~cpp
+
+auto pd = convolution_forward::primitive_desc(opd, attr, eng);
+convolution_forward prim;
+do {
+    prim = convolution_forward(pd, true);
+} while (prim == convolution_forward() && pd.next_impl());
+
+~~~
+
+Pros:
+1. Minimal changes required in documentation;
+2. A slightly simpler control flow compared to [Option
+   1](#option-1-leave-as-is);
+
+Cons:
+1. The user is responsible for ensuring that a valid primitive implementation exists.
+
+### Option 3: fallback to the reference implementation (personal preference)
+
+For this option we introduce `std::unique_ptr<primitive_desc_t> pd_ref_`,  which
+stores the reference implementation for a particular primitive. in
+`dnnl_primitive_desc_iterator` and in `dnnl_primitive_desc`. In case a JIT
+implementation fails during its generation for any reason, the reference
+implementation will be expected to handle any (or most) use cases supported by
+the library, such as attributes, memory layouts, and data types.
+
+In the context of primitive cache, in order to avoid repeatedly failing to
+generate a JIT primitive implementation, we can:
+1. Set `bool primitive_t::is_initialized_` (which is inherited from `struct
+   c_compatible`) to `false` during construction, and change to `true` if
+   `primitive_t::init(engine_t *engine, bool use_global_scratchpad)` returns
+   `dnnl_success`;
+2. Include `bool primitive_t::is_initialized_` into `primitive_hashing::key_t`
+   to compute the primitive hash;
+2. If `is_initialized_ == true`, a valid JIT implementation is found in cache
+   and it will be deployed for computation;
+3. If `is_initialized_ == false`, an invalid JIT implementation is found in
+   cache and the reference implementation will be deployed for computation.
+
+Pros:
+1. No changes are required from the user.
+2. Primitive cache does not need to store the reference implementation since it
+   is lightweight;
+3. Almost unlikely for the application to crash (requires validation);
+
+Cons:
+1. The limitations of the fallback reference implementation have to be
+   documented, such as unsupported options (e.g. lack of reference winograd
+   convolution);
+2. Very poor performance;
+3. Primitive descriptor queries have to take into account the fallback implementation;
+4. Changes required for scratchpad memory tracking to take into account an extra
+   implementation.
+5. Primitive cache stores an (invalid) JIT implementation 
+
+### Option 4: fallback to the next successfull implementation
+
+For this option, we introduce `std::unique_ptr<dnnl::impl::primitive_desc_t>
+pd_fb_`, which stores an additional primitive implementation, in
+`dnnl_primitive_desc_iterator` and in `dnnl_primitive_desc`. Once the first
+valid primitive implementation is assigned to
+`dnnl_primitive_desc_iterator::pd_` when iterating over the implementation list,
+the iterating process continues until `pd_fb_` will also be assigned some
+implementation.
```

Comment:
How will we guarantee that the memory formats of the next (fallback) implementation match the ones of the first (main) implementation? Or you suggest to create the main pd first, and then create a second one based on the queries from the first one?
If so, what would guarantee that the second one won't fail during jitting just like the first one?

Reply:
Looks like there are quite a few issues I didn't think about for this option, thanks Evarist. I agree this option is the least appealing, at least given that any JIT implementation can fail at any point. The only benefit here would be not losing too much performance using `pd_fb_` instead of `pd_`, however it may be simpler to introduce a heuristic inside `pd_` to prevent an implementation being deployed rather than addressing the issues you mentioned.

Reply:
Yeah, agree.

Reply:
Looks like we are going with option 1. Resolving the discussion.